sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/GeneratedColumn.scala

```diff
@@ -162,6 +163,10 @@ object GeneratedColumn {
         s"generation expression data type ${analyzed.dataType.simpleString} " +
         s"is incompatible with column data type ${dataType.simpleString}")
     }
+    if (analyzed.exists(e => SchemaUtils.hasNonDefaultCollatedString(e.dataType))) {
+      throw unsupportedExpressionError(
+        "generation expression cannot contain non-default collated string type")
```

Comment:
It is not fully related to the PR, but I believe we should introduce sub-classes of `UNSUPPORTED_EXPRESSION_GENERATED_COLUMN` instead of the `reason` message paremeter.